File: 129-sum-root-to-leaf-numbers/129-sum-root-to-leaf-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        def dfs(root, current_sum):
            if not root:
                return 0
            current_sum = current_sum*10 + root.val
            if not root.left and not root.right:
                return current_sum
            else:
                return dfs(root.left, current_sum) + dfs(root.right, current_sum)
        return dfs(root, 0)

File: 129-sum-root-to-leaf-numbers/129-sum-root-to-leaf-numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
  int res = 0;

public:
  int sumNumbers(TreeNode *root)
  {
    helper(root, res);
    cout << res;
    return res;
  }
  void helper(TreeNode *node, int curr)
  {
    if (!node)
      return;

    curr = curr * 10 + node->val;
    if (!node->left && !node->right)
    {
      // This is
      res += curr;
      return;
    }

    helper(node->left, curr);

    helper(node->right, curr);
  }
};